### ISSUE TITLE
FSI Digitil Twin - AeroSENSE

### ISSUE THREAD

**Issue by time-trader** [octue/octue-sdk-python#73]
### About me

Octue user,  developing a digital twin for HAWT under [AeroSense project](https://www.researchgate.net/project/Aerosense). 

### Request

FSI Module is the overarching service for running simulations and retrieving results, which should be called by other parts of Aerosense.

**Comment by time-trader** [comment]
### Current state:

Currently FSI Twin is merely a python script that:
<ul>
<li>Generates a list of dicts to be used as an input
<li> Iterates over the inputs.
<ul>
<li>  Launches a local child twin using Runner().run()
<li>  Writes results.output_values json to the drive and adds the file to the manifest , with tags based on the input
</ul>
</ul>

**Comment by thclark** [comment]
This is going to be extremely relevant as it's essentially a use case for the issues solved by #46 which is in progress now.

@time-trader before we finish #46 (it'll be released in 0.1.8, you can track progress in #68) we'll head over to your repositories on gitlab and make sure we fully cover your use case. What extra data, if any, would we need to run the services you're defining there?

**Comment by time-trader** [comment]
@thclark The twins in question are f[oam_2d_twine](https://gitlab.com/windenergie-hsr/aerosense/digital-twin/fsi-twins/foam_2d_twine/-/tree/dev) (dev branch), and [FSI-Simulation](https://gitlab.com/windenergie-hsr/aerosense/digital-twin/fsi-twins/fsi-simulation).
to plot results I run [post-processing_twine](https://gitlab.com/windenergie-hsr/aerosense/digital-twin/utilities/post-processing_twine) separately for now.
All the data (template mesh and foam configuration file) are already there. You would need openFoam installed on your system though. You can replace the foam calls in the app with just randomly assigned output. I should probably include compiled solvers in future?

**Comment by thclark** [comment]
@time-trader I'll dockerise it for you so it'll work cloud-side and so the installation is repeatable.

If you had any particular steps or guides for installing OF that you found useful, drop them here otherwise we'll probably be fine without!

**Comment by time-trader** [comment]
No, I installed OpenFOAM straight from their repo. https://openfoam.org/version/8/

**Comment by thclark** [comment]
[Outdated but a good general guide](https://www.cfdengine.com/blog/how-to-install-openfoam-anywhere-with-docker/).

[OpenFOAM's documentation has relocated since I last used it in 2009ish... but they still have the knack of making it completely unintelligible to anyone anywhere on the human-robot spectrum](https://develop.openfoam.com/Development/openfoam/-/wikis/precompiled/docker)

There are images are on [dockerhub]https://hub.docker.com/search?q=openfoam&type=image) so I can base on those, although the concept of tagging has clearly passed all of the various maintainers completely by :sob:

**Comment by time-trader** [comment]
Careful, because OpenFOAM was forked.. so there is one by [ESI](https://www.openfoam.com/)  and one by[ OpenFOAM Foundation. ](https://openfoam.org/) We used the later. But for basic things they are same.

**Comment by thclark** [comment]
Gotcha, yes I was thinking the OpenFOAM Foundation docker images looked a bit more sensible so relieved that they seem to be a bit more sorted overall.

**Comment by time-trader** [comment]
@thclark  It totally slipped my mind that I actually changed some of the OpenFOAM code to include a rough wall extension to k⁻omega turbulence model: https://github.com/time-trader/OpenFOAM-8/tree/master/src/MomentumTransportModels/momentumTransportModels

But I guess I can just `wmake` from new source once the docker container is up?

**Comment by thclark** [comment]
So docker doesn't work quite like that - the idea is that you build an image with everything pre-compiled and then the container is 'stateless', a glorious sounding concept that arises from the fact that any computer in the cloud can crap out at any time for no reason, and basically means "don't store important stuff in the container".

Compiling once a container is running would be a violation of this, because a) it's a long process (so boot time would be intense, it has to be redone on every boot) and b) it'd change the way the container behaves in response to inbound traffic (i.e. changes the state of the container, making it stateful).

Computer science lesson over. You can ignore the above if not interested.

Practically speaking, what we'll do is build an image with your compiled version of OpenFOAM in it, so basically we'll put your `wmake` commands in a dockerfile and away we go!

Can you paste the compilation commands etc that you used (from your terminal history, or if you've a script)?

**Comment by time-trader** [comment]
I just `cd  /openfoam8/src/MomentumTransportModels/momentumTransportModels/`
and then `wmake`
it makes new lib `libmomentumTransportModels.so` at my user folder.

**Comment by thclark** [comment]
And that doesn't need to be linked to openfoam? Or is is replacing an existing, rather than adding a new, transport model?

**Comment by thclark** [comment]
> And that doesn't need to be linked to openfoam? Or is is replacing an existing, rather than adding a new, transport model?

(in which case switching the DLL with the same API is sufficient)

**Comment by time-trader** [comment]
openfoam looks for custom build libs in a specific user folder. Yeah I think we can just replace the dll

**Comment by cortadocodes** [comment]
Enabled by #79 (released in #68)

**Comment by thclark** [comment]
@time-trader that means the SDK is ready to support the parent-child communication, but this issue is a bit more wide ranging than that since it covers openfoam and helping to get you set up with the new communication method, so we're not closing just yet.

However, we're prioritising file management aspects in order to get the aerosense data ingress up and running right now, so will come back to this in a couple of weeks when the ingress is running smoothly.
